packages/@react-aria/utils/docs/useObjectRef.mdx

```diff
@@ -0,0 +1,135 @@
+<!-- Copyright 2021 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License. -->
+
+import {Layout} from '@react-spectrum/docs';
+export default Layout;
+
+import docs from 'docs:@react-aria/utils';
+import {HeaderInfo, FunctionAPI} from '@react-spectrum/docs';
+import packageData from '@react-aria/utils/package.json';
+
+```tsx import
+import {useEffect} from 'react';
+import {useButton} from '@react-aria/button';
+import {useTextField} from '@react-aria/textfield';
+import {useObjectRef} from '@react-aria/utils';
+```
+
+---
+category: Utilities
+keywords: [refs]
+---
+
+# useObjectRef
+
+<p>
+Offers an object ref for a given callback ref or an object ref. Especially
+helfpul when passing forwarded refs (created using <code>React.forwardRef</code>)
+to React Aria Hooks.
+</p>
+
+<HeaderInfo
+  packageData={packageData}
+  componentNames={['useObjectRef']} />
+
+## API
+
+<FunctionAPI function={docs.exports.useObjectRef} links={docs.links} />
+
+## Introduction
+
+The `useObjectRef` hook creates an object ref for a given [ref](https://reactjs.org/docs/refs-and-the-dom.html).
+The given ref may be an object ref or a [callback ref](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs).
+The created ref is particularly helpful when defining a custom component using
+[`React.forwardRef`](https://reactjs.org/docs/forwarding-refs.html), and wanting to pass the forwarded ref to
+a `@react-aria` Hook that accepts a ref (e.g. [`useButton`](useButton.html#example)).
+
+## Example using an object ref
+
+```tsx example
+const TextField = React.forwardRef((props, forwardedRef) => {
+  const {label} = props;
+  const internalRef = React.useRef();
+  const ref = useObjectRef(forwardedRef || internalRef);
```

Comment:
Just wanna make sure I understand the code here, but is `internalRef` here necessary? `useObjectRef` will return an object ref if `forwardedRef` is undef right?

Reply:
Yes, that is correct -- `useObjectRef` will always return a `ref` object. I guess the example was little bit too contrived based on https://github.com/adobe/react-spectrum/issues/834#issuecomment-786927889.

Removed to avoid the confusion :)

Also, I think for this example to be clearer, it would make more sense to make it more TypeScript.

Refactored that as well.

Reply:
Forgot that docs won't take TypeScript :/ I had to revert. Perhaps I can try adding a TypeScript example that won't necessarily execute..? Or would that also not be possible?

Reply:
Might be worth attaching a codesandbox example to the docs then, kinda like https://react-spectrum.adobe.com/react-aria/useComboBox.html#styled-examples

Reply:
That sounds good 👍🏼  Shall I do that after merging this PR so I can pull the `useObjectRef` util on codesandbox? Or how do we wanna do that?

Reply:
oh yeah, probably can just do a followup update after. In fact, you'll probably need to separate the docs from this PR actually so it doesn't get pushed up to the docs site immediately on merge

Reply:
Ah, yeah, you're right 👍🏼 
Done.